include/dali/core/bitmask.h

```diff
@@ -0,0 +1,231 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_BITMASK_H_
+#define DALI_CORE_BITMASK_H_
+
+#include <vector>
+#include <cassert>
+#include "dali/core/util.h"
+
+namespace dali {
+
+/**
+ * @brief A vector of bits with a utility for quickly searching for set/cleared bits.
+ */
+struct bitmask {
+ public:
+  using bit_storage_t = uint64_t;
+
+  static constexpr const int storage_bits = sizeof(bit_storage_t) * 8;
+  static constexpr const int storage_bits_log = ilog2(storage_bits);
+
+  /**
+   * @brief Find the first bit with given value, starting at start
+   *
+   * @return The index of the first bit found of the size of the bit mask if not found.
+   */
+  ptrdiff_t find(bool value, ptrdiff_t start = 0) const {
+    assert(start >= 0);
+    if (start >= size_)
+      return size_;
+    auto index = start >> storage_bits_log;
+    int bit = start & (storage_bits - 1);
+    bit_storage_t flip = 0;
+    if (!value)
+      flip = ~flip;
+
+    // we're not starting at a storage word boundary - look at the first word first
+    if (bit > 0) {
+      auto word = storage_[index] ^ flip;
+      word >>= bit;  // skip bits before start
+      if (word)
+        return start + ctz(word);
+      index++;
+    }
+
+    ptrdiff_t num_words = storage_.size();
+    for (; index < num_words; index++) {
+      auto word = storage_[index] ^ flip;
+      if (word) {  // we've found something...
+        return (index << storage_bits_log) | ctz(word);
+      }
+    }
+
+    return size_;
+  }
+
+  /**
+   * @brief Fill a range of bits with given value
+   */
+  void fill(ptrdiff_t start, ptrdiff_t end, bool value) {
+    if (start >= end)
+      return;
+    ptrdiff_t start_idx = start >> storage_bits_log;
+    ptrdiff_t end_idx = end >> storage_bits_log;
+    bit_storage_t ones = ~bit_storage_t(0);
+    bit_storage_t start_mask = ones;
+    bit_storage_t end_mask = 0;
+    if (start & (storage_bits - 1)) {
+        start_mask <<= (start & (storage_bits - 1));
+    }
+    if (end & (storage_bits - 1)) {
+        end_mask = ones >> (storage_bits - (end & (storage_bits - 1)));
+    }
+    if (start_idx == end_idx) {
+        start_mask &= end_mask;
+    }
+    ptrdiff_t idx = start_idx;
+    if (value)
+        storage_[idx++] |= start_mask;
+    else
+        storage_[idx++] &= ~start_mask;
+    if (value) {
+        for (; idx < end_idx; idx++)
+            storage_[idx] = ones;
+    } else {
+        for (; idx < end_idx; idx++)
+            storage_[idx] = 0;
+    }
+    if (end_mask && start_idx != end_idx) {
+      if (value)
+        storage_[end_idx] |= end_mask;
+      else
+        storage_[end_idx] &= ~end_mask;
+    }
+  }
+
+  /**
+   * @brief Fill the entire mask with given value.
+   */
+  void fill(bool value) {
+    size_t s = size();
+    clear();
+    resize(s, value);
+  }
+
+  /**
+   * @brief A pseudoreference to a single bit in the mask
+   */
+  struct bitref {
```

Comment:
Note to reviewers: there are no iterators and that's intentional. There are few algorithms that can be used on a collection of booleans and those taken from STL would perform poorly.